image/generate.go

```diff
@@ -0,0 +1,362 @@
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+const (
+	refTag = "latest"
+
+	layoutStr = `{"imageLayoutVersion": "1.0.0"}`
+
+	configStr = `{
+    "created": "2015-10-31T22:22:56.015925234Z",
+    "author": "Alyssa P. Hacker <alyspdev@example.com>",
+    "architecture": "amd64",
+    "os": "linux",
+    "config": {
+        "User": "alice",
+        "ExposedPorts": {
+            "8080/tcp": {}
+        },
+        "Env": [
+            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+            "FOO=oci_is_a",
+            "BAR=well_written_spec"
+        ],
+        "Entrypoint": [
+            "/bin/my-app-binary"
+        ],
+        "Cmd": [
+            "--foreground",
+            "--config",
+            "/etc/my-app.d/default.cfg"
+        ],
+        "Volumes": {
+            "/var/job-result-data": {},
+            "/var/log/my-app-logs": {}
+        },
+        "WorkingDir": "/home/alice"
+    },
+    "rootfs": {
+      "diff_ids": [
+        "sha256:c6f988f4874bb0add23a778f753c65efe992244e148a1d2ec2a8b664fb66bbd1",
+        "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef"
+      ],
+      "type": "layers"
+    },
+    "history": [
+      {
+        "created": "2015-10-31T22:22:54.690851953Z",
+        "created_by": "/bin/sh -c #(nop) ADD file:a3bc1e842b69636f9df5256c49c5374fb4eef1e281fe3f282c65fb853ee171c5 in /"
+      },
+      {
+        "created": "2015-10-31T22:22:55.613815829Z",
+        "created_by": "/bin/sh -c #(nop) CMD [\"sh\"]",
+        "empty_layer": true
+      }
+    ]
+}
+`
+)
+
+var (
+	refStr = `{"digest":"<manifest_digest>","mediaType":"application/vnd.oci.image.manifest.v1+json","size":<manifest_size>}`
+
+	manifestStr = `{
+    "annotations": null,
+    "config": {
+        "digest": "<config_digest>",
+        "mediaType": "application/vnd.oci.image.config.v1+json",
+        "size": <config_size>
+    },
+    "layers": [
+        {
+            "digest": "<layer_digest>",
+            "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
+            "size": <layer_size>
+        }
+    ],
+    "mediaType": "application/vnd.oci.image.manifest.v1+json",
+    "schemaVersion": 2
+}
+ `
+)
+
+type tarContent struct {
+	header *tar.Header
+	b      []byte
+}
+
+type imageLayout struct {
+	rootDir  string
+	layout   string
+	ref      string
+	manifest string
+	config   string
+	tarList  []tarContent
+}
+
+func createImageLayoutBundle(il imageLayout) error {
+	err := os.MkdirAll(filepath.Join(il.rootDir, "blobs", "sha256"), 0700)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(filepath.Join(il.rootDir, "refs"), 0700)
+	if err != nil {
+		return err
+	}
+
+	// create image layout file
+	err = createLayoutFile(il.rootDir)
+	if err != nil {
+		return err
+	}
+
+	// create image layer blob file.
```

Comment:
I'd prefer this function be public, [create whatever boilerplate the spec requires image layouts to have, and then stop][1].  Filling in example layers, JSON blobs, and refs isn't useful (we can always distribute the example image directly if that content is useful).  What I think users want (among other things) is a tool to create a new, *empty* image layout so they can start populating it with their content.  And that's what #5's [`oci-image-init image-layout …`][2], [`CreateDir`][3], and [`CreateTarFile`][4] are doing.

[1]: https://github.com/opencontainers/image-tools/pull/101#discussion_r93663062
[2]: https://github.com/opencontainers/image-tools/pull/5/files#diff-0cf1d5acd82a85146cf1eaf17df69bbcR10
[3]: https://github.com/opencontainers/image-tools/pull/5/files#diff-1d84590cedd145294b2105cc459ef5f3R43
[4]: https://github.com/opencontainers/image-tools/pull/5/files#diff-745d1b74657ed8e87ac8c4e3f520f73aR212